### DISCUSSION TITLE
dotenv related issue

### DISCUSSION THREAD

**Discussion by Lalrinfela** [quasarframework/quasar#9719] [General - Components / Directives / etc] [answered]
When using dotenv npm packages, and as per the instructions here: https://next.quasar.dev/quasar-cli/handling-process-env

The issue is that the env file are read only once during the time of dev/build. If the env files are updated, it do not updated in the process.env value until the whole 'quasar dev' or 'quasar build' is rerun.

Am I doing something wrong or is this the expected behaviour. It's not what i expected.

**Comment by MilosPaunovic** [comment]
Hello @Lalrinfela,

I've been using `Quasar 1.x` for some time now with `dotenv`, and I think that reloading you app is required if you change the environment variables, because the import of 'em is in the `build` attribute of `quasar.conf.js` according to [this](https://next.quasar.dev/quasar-cli/handling-process-env#adding-to-process-env)

```js
// quasar.conf.js

module.exports = function (ctx) {
  return {
    // ...

    build: {
      // passing down to UI code from quasar.conf.js
      env: {
        API: ctx.dev ? 'https://dev.api.com' : 'https://prod.api.com'
      }
    }
  }
}
```
So it just makes sense that you have to reload the app for the system to acknowledge changes.

**Reply by Lalrinfela** [reply]
I do understand this. But when quasar are deployed, we deploy content of the dist folder. So if the .env data are bind by the 'quasar dev' or 'quasar build', then it makes no sense to me. Why do use it even in the first place when we can just put the content of the env in any other file as well.

The need to run 'quasar build' in the production server with the production .env files is cumbersome. It's that i deploy the dist folder and use a different .env files in the server and the whole apps should works just fine.

Like PHP Laravel framework is what i was thinking the dotenv to work but it seems its not.

I have been spending the past 4 hours as to make it work dynamically without the need to compile. But, i come up with my own way of implementation.

Thanks for the response. :-)

**Reply by MilosPaunovic** [reply]
Vue build process states this [here](https://cli.vuejs.org/guide/mode-and-env.html#environment-variables):

`Note that only NODE_ENV, BASE_URL, and variables that start with VUE_APP_ will be statically embedded into the client bundle with webpack.DefinePlugin. It is to avoid accidentally exposing a private key on the machine that could have the same name.`

Meaning your environment variables can stay private. I understand what would you like to accomplish and it does make a lot of sense to me as well, it's just that the Node and PHP processes are different...

My suggestion is to create a rudimentary GitHub Action (or its equivalent on the system you use) which will run the build for you; it saves time and gives you opportunity to always know which version of code is deployed.

**Comment by hawkeye64** [comment, marked as answer]
@Lalrinfela env vars get "baked" into the output files at build time. This is not available for run-time. If you need run-time, then your server should serve a json file. Be aware, with this approach, anyone can open dev tools and see the contents of your json file. Quasar obfuscates your `process.env` data from prying eyes.

**Reply by Lalrinfela** [reply]
Thanks @hawkeye64 

My approach now it that i still baked the custom env data in the process.env. But, this time, i just detect the hostname and apply different settings.

It would be great if we could just use a .env file dynamically and securely like PHP as mentioned before. Maybe it's beyond my understanding.

Specially my case is a SaaS where i have to redeploy sites in sub-domain with specific sub-domain settings.

Love Quasar. You guys are awesome.!
